Calling rest client from message listener causes ClassNotFoundException in development mode

If rest client is called from JMS message listener (class annotated with `ResourceEndpoint`), one gets ClassNotFoundException. E.g:

```
@ResourceEndpoint(activationSpecConfigKey = "myqueue") 
public class MyResourceEndpoint implements jakarta.jms.MessageListener {
    @Inject
    @RestClient
    GreetingRestClient restClient;

    @Override
    public void onMessage(Message message) {
        try {
            System.out.println("XXXXXXXXXXXXXXXXXXXXXXXXXXXXX  "+message.getBody(String.class));

            try {
                var o = restClient.greetUser("Hello");
            } catch (Exception e) {
                /*
                 * In dev mode (quarkus:dev) one gets
                 * Caused by: java.lang.ClassNotFoundException: services.api.GreetingRestClient
                 * at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:542)
                 * at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:517)
                 * at java.base/java.lang.Class.forName0(Native Method)
                 * at java.base/java.lang.Class.forName(Class.java:467)
                 * at services.api.GreetingRestClient$$CDIWrapper.<init>(Unknown Source)
                 * at services.api.GreetingRestClient$$CDIWrapper_Bean.doCreate(Unknown Source)
                 *
                 * Production mode works fine
                 */
                e.printStackTrace();
            }

        } catch (JMSException e) {
            throw new RuntimeException(e);
        }
    }
}
```
Moving the rest call to CDI bean doesn't help. It looks like the `GreetingRestClient` is being loaded with the `Quarkus Base Runtime ClassLoader` instead of `Quarkus Runtime ClassLoader`, which causes the exception.

I created small project to reproduce the issue [https://github.com/mkomadel/quarkus-iron-jamacar-artemis/tree/main](https://github.com/mkomadel/quarkus-iron-jamacar-artemis/tree/main ) . It uses quarkus-artemis-jms-ra, which in turn uses iron-jamacar extension. 

Steps to reproduce:

1. clone project from the above repository `git clone https://github.com/mkomadel/quarkus-iron-jamacar-artemis.git`
2.  start apache MQ Artemis locally on port 61616 (docker example `docker run --detach --rm -p 8161:8161 -p 61616:61616 -p 5672:5672 -e BROKER_IP="0.0.0.0" -e AMQ_EXTRA_ARGS="--relax-jolokia" -e AMQ_USER=amquser -e AMQ_PASSWORD=amquser --name artemis quay.io/artemiscloud/activemq-artemis-broker:1.0.28`)
3. start quarkus service in dev mode: `mvn quarkus:dev`
4. send message to queue `curl http://localhost:8080/hello`
5. observe ClassNotFoundException in stdout

--- a/integration-tests/artemis-dev-classloader/src/main/java/io/quarkiverse/ironjacamar/classloader/MyMessageListener.java
+++ b/integration-tests/artemis-dev-classloader/src/main/java/io/quarkiverse/ironjacamar/classloader/MyMessageListener.java
@@ -28,8 +28,7 @@ public class MyMessageListener implements MessageListener {
         try (JMSContext context = connectionFactory.createContext()) {
             JMSProducer producer = context.createProducer();
             Queue queue = context.createQueue("reply");
-            // producer.send(queue, myRestClient.get().hello());
-            producer.send(queue, "Hello, World!");
+            producer.send(queue, myRestClient.get().hello());
         }
     }
 }

--- a/runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/ClassLoaderMessageEndpoint.java
+++ b/runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/ClassLoaderMessageEndpoint.java
@@ -1,0 +1,39 @@
+package io.quarkiverse.ironjacamar.runtime.endpoint;
+
+import java.lang.reflect.Method;
+
+import jakarta.resource.ResourceException;
+import jakarta.resource.spi.endpoint.MessageEndpoint;
+
+/**
+ * A {@link MessageEndpoint} that sets the context class loader to a specific class loader before calling the delegate
+ */
+class ClassLoaderMessageEndpoint extends MessageEndpointWrapper implements MessageEndpoint {
+    private final ClassLoader classLoader;
+
+    private ClassLoader originalClassLoader;
+
+    public ClassLoaderMessageEndpoint(MessageEndpoint delegate, ClassLoader classLoader) {
+        super(delegate);
+        this.classLoader = classLoader;
+    }
+
+    @Override
+    public void beforeDelivery(Method method) throws NoSuchMethodException, ResourceException {
+        this.originalClassLoader = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(classLoader);
+        super.beforeDelivery(method);
+    }
+
+    @Override
+    public void afterDelivery() throws ResourceException {
+        Thread.currentThread().setContextClassLoader(originalClassLoader);
+        super.afterDelivery();
+    }
+
+    @Override
+    public void release() {
+        this.originalClassLoader = null;
+        super.release();
+    }
+}

--- a/runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/ClassLoaderMessageEndpoint.java
+++ b/runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/ClassLoaderMessageEndpoint.java
@@ -8,7 +8,7 @@ import jakarta.resource.spi.endpoint.MessageEndpoint;
 /**
  * A {@link MessageEndpoint} that sets the context class loader to a specific class loader before calling the delegate
  */
-class ClassLoaderMessageEndpoint extends MessageEndpointWrapper implements MessageEndpoint {
+class ClassLoaderMessageEndpoint extends MessageEndpointWrapper {
     private final ClassLoader classLoader;
 
     private ClassLoader originalClassLoader;

--- a/runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/DefaultMessageEndpointFactory.java
+++ b/runtime/src/main/java/io/quarkiverse/ironjacamar/runtime/endpoint/DefaultMessageEndpointFactory.java
@@ -14,6 +14,7 @@ import io.quarkiverse.ironjacamar.Defaults;
 import io.quarkiverse.ironjacamar.ResourceAdapterFactory;
 import io.quarkus.arc.Arc;
 import io.quarkus.arc.ArcContainer;
+import io.quarkus.runtime.LaunchMode;
 import io.smallrye.common.annotation.Identifier;
 
 /**
@@ -24,6 +25,7 @@ public class DefaultMessageEndpointFactory implements MessageEndpointFactory {
     private final Class<?> endpointClass;
     private final String identifier;
     private final ResourceAdapterFactory resourceAdapterSupport;
+    private final ClassLoader classLoader;
     private Boolean transacted;
 
     /**
@@ -37,6 +39,7 @@ public class DefaultMessageEndpointFactory implements MessageEndpointFactory {
         this.endpointClass = endpointClass;
         this.identifier = identifier;
         this.resourceAdapterSupport = adapterFactory;
+        this.classLoader = Thread.currentThread().getContextClassLoader();
     }
 
     @Override
@@ -61,12 +64,16 @@ public class DefaultMessageEndpointFactory implements MessageEndpointFactory {
     }
 
     @Override
-    public MessageEndpoint createEndpoint(XAResource xaResource) throws UnavailableException {
-        final MessageEndpoint endpoint;
+    public MessageEndpoint createEndpoint(XAResource xaResource) {
+        MessageEndpoint endpoint;
         if (xaResource == null) {
             endpoint = NoopMessageEndpoint.INSTANCE;
         } else {
             endpoint = new TransactionAwareMessageEndpoint(xaResource);
+        }
+        // When running in dev mode, we don't want to wrap the endpoint
+        if (LaunchMode.current() != LaunchMode.NORMAL) {
+            endpoint = new ClassLoaderMessageEndpoint(endpoint, classLoader);
         }
         return resourceAdapterSupport.wrap(endpoint, getEndpointInstance());
     }

